### DISCUSSION TITLE
Help an idiot make an extremely simple batch convert script and get it working

### DISCUSSION THREAD

**Discussion by MintJamsLP** [ImageMagick/ImageMagick#4993] [Help]
I am looking to make a *imconvert*.bat script (imconvert just a made up file name) that will batch convert images from the command prompt. Ideally I would like to be able to type;

*imconvert folderpath* (for a specific folder)

or just

*imconvert* to convert all images within current folder

I need to be able to convert all images in a folder to PNG. Currently most files are WEBP, but some JPEG.

If it can handle files within subfolders also, that would be better.

*convert input.webp output.png* Is what I found within the official documentation. Some examples also showed *convert image.jpg*

As well, how am I able to alter PATH / Command Prompt settings so that;

When I open prompt from a specific folder's address bar, the files are read from, and converted images are saved to that same folder?

While ImageMagick is in its own folder on another drive.

I did this with yt-dlp following some instructions, but I can't figure out how to do it with ImageMagick.

Thank you for any help!

**Comment by snibgo** [comment]
What version of IM? What is your platform, eg bash or Windows 8.1 or what?

I suggest you ensure that the IM directory is on your system path.

To convert all WEBP files to PNG in a folder, I suggest using "magick mogrify" for IM v7,or "mogrify" for IM v6.
```
magick mogrify -format PNG *.webp
```
If the image files are not in your current directory, then give their path:
```
magick mogrify -format PNG mysubdir\*.webp
```
For the path, in a script you can use %1 for the first parameter to the script.

**Reply by MintJamsLP** [reply]
Hello, thanks for your help. I have read up a lot, and with trial and error I am getting there, but still a little unsure of the script aspect.

IM 7.1.0, using Win10 x64, build 19044.

I figured out how to put the IM directory into the system path (X:\ImageMagick\). While I was in the Environment Variables, I saw the addition I had made for yt-dlp, so I was able to use that as an example for IM. 

If I am not mistaken, this should allow me to use convert.exe, mogrify.exe etc (all within IM directory) in the command prompt, within any folder. Great!

    magick mogrify -format PNG *.webp

When using this, I could not get it to work. It gave me this result:

    mogrify: UnableToOpenBlob 'PNG:': Invalid argument @ error/blob.c/OpenBlob/3527.
    mogrify: UnrecognizedImageFormat `PNG' @ error/mogrify.c/MogrifyImageCommand/5132.


I uninstalled IM and reinstalled, made sure that the IM directory was still in the path, and retried 

    magick mogrify -format PNG *.webp

It worked! Thank you very much.

I modified it to 

    magick mogrify -path X:\IMPNG -format PNG *.jpg -format PNG .webp

This of course converts both jpg and webp at the same time, and deposits the PNG into X:\IMPNG

Is there a command to convert *all* image types to PNG without manually naming each type?

As for the script...

Would I create a .bat file (IMPNG.bat) with this code;

    magick mogrify -path X:\IMPNG -format PNG *.jpg -format PNG .webp

and then save it within the ImageMagick root folder? (same location as mogrify.exe)

Would this allow me to open CMD from the .webp's (etc) folder, type "IMPNG" into the command line, and have mogrify run as before?


Thank you again for all your help!

**Comment by snibgo** [comment]
> Is there a command to convert all image types to PNG without manually naming each type?

If _all_ the files in the input directory are image files, then you can simply use * as the filename. If any of the files are not image files, IM will fail to read them.

You want all output formats to be PNG, so there is no need to repeat "-format PNG".
```
magick mogrify -path X:\IMPNG -format PNG *.jpg *.webp
```
If your current directory is X:\IMPNG, the "-path" option is redundant.

> ... and then save it within the ImageMagick root folder? (same location as mogrify.exe)

You could, but I suggest not. Instead, I suggest putting your scripts in a different folder. Add that to your system path, if you want.

**Reply by MintJamsLP** [reply]
Okay, this is great! 

I have the script written and stored in a separate folder, which is added to the directory for ease of use. (IMPNG.bat)

After testing it on some images, it's now up to:

    magick mogrify -path X:\IMPNG -format PNG *.jpg *.webp *.png *.jpeg *.avif

I had to use both .jpeg and .jpg to read both extensions, and .png is so that files already in PNG format get moved to X:\IMPNG also.

It is doing *almost* all of what I am looking for. 

I am able to type "cmd /k IMPNG" into the address bar of any folder and it goes to work on that folder. It dumps those images into X:\IMPNG for manual sorting into a new folder.

Is there a way to delete the original files after conversion with the script? I do understand the risk of something going wrong, but this would only be used on folders that are backed up. That way I can (remove -path X:\IMPNG and) keep the original folder name and subfolder structure for the converted images, so I don't have to manually re-folder. (I looked over the documentation for deletion commands but didn't see anything)

or;

Is is possible for magick mogrify to read images within multiple subfolders in one operation, and retain that folder structure when outputting to X:\IMPNG ?  (from what I can see, I don't *think* so, but I am unsure)


Also, although everything appears to be converting fine, there are a couple of issues.

This presents after the operation completes, despite working;

    mogrify: unable to open image '*.png': Invalid argument @ error/blob.c/OpenBlob/3527.
    mogrify: unable to open image '*.jpeg': Invalid argument @ error/blob.c/OpenBlob/3527.
    mogrify: unable to open image '*.jpeg': Invalid argument @ error/blob.c/OpenBlob/3527.
    mogrify: unable to open image '*.avif': Invalid argument @ error/blob.c/OpenBlob/3527.
    mogrify: unable to open image '*.avif': Invalid argument @ error/blob.c/OpenBlob/3527.


Also, when presented with two files with the same name, but different extensions, it seems only one file is converted. That's fine, as long as it is the larger file. I suppose to achieve that I should reorder my script from larger to smaller extension type;

    magick mogrify -path X:\IMPNG -format PNG *.png *.jpg *.jpeg *.webp *.jpeg *.avif

**Comment by snibgo** [comment]
> Is there a way to delete the original files after conversion with the script?

Not within IM, no. IM used to have a feature named "ephemeral:", but that has been removed. You could have "del" commands in the script to remove the input files. Or even delete the directory ("rd" command), and rename the output directory to the original directory name.

> Is is possible for magick mogrify to read images within multiple subfolders in one operation, and retain that folder structure when outputting to X:\IMPNG ?

Again, not within IM. A shell "for" loop can loop through subdirectories.

> mogrify: unable to open image '*.png':

That message will be issued if you have no files that match.

> Also, when presented with two files with the same name, but different extensions, it seems only one file is converted.

I doubt that. I expect both files are converted, but both outputs will have the same name, so the second will overwrite the first. IM has no feature to prevent the overwriting.

**Comment by henryreith** [comment]
Just a quick thanks, this discussion has helped me out a lot.
